Google.Api.Gax.Grpc.Tests/Rest/PartialDecodingStreamReaderTest.cs

```diff
@@ -0,0 +1,165 @@
+            Assert.False(result);
+
+            result = await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false);
+            Assert.False(result);
+        }
+
+        /// <summary>
+        /// Test when data breaks off unexpectedly.
+        /// </summary>
+        [Fact]
+        public async void DecodingStreamReaderTestIncomplete()
+        {
+            var s = DataStr.Split(',')[0]; // right after the first element ends
+            StreamReader reader = new MyStreamReader(s.Split('\n'));
+            var decodingReader = new PartialDecodingStreamReader<JObject>(Task.FromResult(reader), JObject.Parse);
+
+            var result = await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false);
+            Assert.True(result);
+            Assert.NotNull(decodingReader.Current);
+            Assert.Equal(decodingReader.Current["foo"], 1);
+
+            var ex = await Assert.ThrowsAsync<InvalidOperationException>(async () =>
+                await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false));
+            Assert.Contains("Closing `]` bracket not received after iterating through the stream.", ex.Message);
+        }
+
+        /// <summary>
+        /// Test when data is empty array.
+        /// </summary>
+        [Fact]
+        public async void DecodingStreamReaderTestEmpty()
+        {
+            StreamReader reader = new MyStreamReader(new[] {"[]"});
+            var decodingReader = new PartialDecodingStreamReader<JObject>(Task.FromResult(reader), JObject.Parse);
+
+            var result = await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false);
+            Assert.False(result);
+
+            result = await decodingReader.MoveNext(CancellationToken.None).ConfigureAwait(false);
+            Assert.False(result);
+        }
+    }
+
+    /// <summary>
+    /// A fake of a StreamReader emitting given strings
+    /// </summary>
+    internal class MyStreamReader : StreamReader
+    {
+        private readonly Queue<string> _queue;
+
+        /// <summary>
+        /// Cannot override EndOfStream, so have to nudge
+        /// the base class to do this.
+        /// Initialize it with a non-empty stream and later read
+        /// that one to end.
+        /// </summary>
+        /// <param name="strings"></param>
+        public MyStreamReader(IEnumerable<string> strings) : base(new MemoryStream(new byte[1]))
+        {
+            _queue = new Queue<string>(strings);
+        }
+
+        public override Task<int> ReadAsync(char[] buffer, int index, int count)
+        {
+            if (_queue.Count <= 0)
+            {
+                base.ReadToEnd(); // EndOfStream starts to return true
+                return Task.FromResult(0);
+            }
+
+            var nextString = _queue.Dequeue();
+            for (int i = 0; i < nextString.Length; i++)
```

Comment:
This is ignoring both `index` and `count`.
If we're confident that `count` will always be big enough, let's assert that and then just use `nextString.Length`, but we should definitely observe `index`. I'm somewhat surprised this is working without that.

Reply:
👍